Remove middle structs

ali-bahjati opened the pull request:
Before converting VAA payloads to PriceInfo we convert them to PriceAttestation which is the equivalent of the payload wire format. However, this middle conversion has some costs, because (1) we copy stuff twice, and (2) we have some unused fields. 

This PR removes the middle structs and creates PriceInfo directly from the VAA payload. 

The gas improvement is not substantial, but still good (~14k gas save)

The gas snapshot difference:
```
testBenchmarkGetUpdateFee() (gas: -22 (-0.016%)) 
testBenchmarkUpdatePriceFeedsIfNecessaryNotFresh() (gas: 110 (0.054%)) 
testBenchmarkUpdatePriceFeedsIfNecessaryFresh() (gas: -22212 (-5.063%)) 
testBenchmarkUpdatePriceFeedsFresh() (gas: -22272 (-5.374%)) 
testBenchmarkUpdatePriceFeedsNotFresh() (gas: -22236 (-5.736%)) 
```

jayantk commented on ethereum/test/pyth.js:
```diff
@@ -298,45 +294,27 @@ contract("Pyth", function () {
+            assert.equal(parsed[i].priceInfo.emaPrice.conf, 42);
+
+            assert.equal(parsed[i].priceInfo.emaPrice.publishTime, publishTime);
+            assert.equal(parsed[i].priceInfo.price.publishTime, publishTime);
```
can you also add a test here for the case where the status is unknown?

ali-bahjati replied:
There is already a test for it. It didn't need anychange as it uses end to end just like this one. 

https://github.com/pyth-network/pyth-crosschain/blob/main/ethereum/test/pyth.js#L695
I had added this when I added this logic and it's a good kind of test that needs no change when you change internals  :D

jayantk commented on ethereum/contracts/pyth/PythInternalStructs.sol:
```diff
@@ -62,17 +33,4 @@ contract PythInternalStructs {
-        HALTED: The price feed is not currently updating because trading in the product has been halted.
-        AUCTION: The price feed is not currently updating because an auction is setting the price.
-    */
-    enum PriceAttestationStatus {
```
i take it the enum costs extra gas ? it's a lot nicer to have the enum than a number so would be nice to keep this if possible.

ali-bahjati replied:
Probably there is a gas cost, maybe not much. But I rather keep it out as we soon will remove status and previous fields from price attestation (it will also save gas + makes logic simpler)

jayantk commented on ethereum/contracts/pyth/PythInternalStructs.sol:
```diff
@@ -9,38 +9,9 @@ import "@pythnetwork/pyth-sdk-solidity/PythStructs.sol";
 contract PythInternalStructs {
     using BytesLib for bytes;
 
-    struct BatchPriceAttestation {
```
it's safe to delete these structs because they're never stored anywhere or shared with users outside the contract?

ali-bahjati replied:
yes

ali-bahjati commented:
By @njkumr suggestion, I moved the processing part in the parsing method. This will reduce a lot of gas because we won't append to memory any more. 

Difference:

```
testBenchmarkGetUpdateFee() (gas: -22 (-0.016%)) 
testBenchmarkUpdatePriceFeedsIfNecessaryNotFresh() (gas: 110 (0.054%)) 
testBenchmarkUpdatePriceFeedsIfNecessaryFresh() (gas: -7455 (-1.758%)) 
testBenchmarkUpdatePriceFeedsFresh() (gas: -7532 (-1.884%)) 
testBenchmarkUpdatePriceFeedsNotFresh() (gas: -7300 (-1.959%)) 
Overall gas change: -22199 (-5.564%)
```

I updated the gas snapshot above as well.

jayantk commented on ethereum/test/pyth.js:
```diff
@@ -283,61 +286,41 @@ contract("Pyth", function () {
+
+            const price = await this.pythProxy.getPriceUnsafe(price_id);
+            assert.equal(price.price, priceVal.toString());
+            assert.equal(price.conf, "101"); // Fixed in the fixture.
```
i don't understand the code comment here

jayantk commented on ethereum/test/pyth.js:
```diff
@@ -283,61 +286,41 @@ contract("Pyth", function () {
+            price: "1337",
+        });
+
+        // Then prices should be available because the valid period is now 120 seconds
```
can you add a test for the case where the status is unknown?

jayantk reviewed (commented):
generally fine, but would like to see a test for the case where status=unknown. also had some questions about how the safety of the migration (i think this is mostly for my own understanding).

njkumr reviewed (approved):
the changes lgtm!
